<a href="https://colab.research.google.com/github/Somaskandan931/flyrank-ml-2026-Somaskandan931/blob/main/work/notebooks/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/flyrank-bih/flyrank-ml-internship-starter/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

# 1. My Lane

## Chosen Lane
Refresh / Content Opportunity Scoring

## Why I chose this lane

I chose the Refresh / Content Opportunity Scoring lane because it directly addresses a practical business decision: given limited time and resources, which content pages should be reviewed first?

The goal is not to predict whether updating a page will definitely improve traffic. Instead, the goal is to prioritize pages that show evidence of decline or opportunity so that human reviewers can spend their effort where it is most valuable.

This lane matches the internship's focus on producing ranked action recommendations using observable search and engagement signals rather than reproducing existing product rules.

## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

## Research Question

### Search Question
Which content pages should be prioritized for manual review based on observed search and engagement signals?

### Decision
Decide which content pages should be reviewed first by the SEO or content team.

### Unit of Analysis
One content page.

### Output
A ranked priority score that orders pages from highest to lowest review priority.

### Action
The reviewer examines the highest-ranked pages and decides whether the content should be refreshed, rewritten, expanded, or left unchanged.

### Cost of a Wrong Recommendation

**False Positive:** A page is recommended for review even though it does not need attention, resulting in wasted reviewer time and effort.

**False Negative:** A page that genuinely needs review is missed, allowing declining content to continue losing search visibility and traffic.

### Why Machine Learning Can Help
The dataset contains many pages and multiple search and engagement signals. Reviewing every page manually is difficult, so machine learning can help prioritize pages for human review based on observed historical patterns.

## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [ ]:
# ============================================================
# Step 1: Clone the GitHub repository (only once)
# ============================================================

import os
import pandas as pd

REPO_NAME = "flyrank-ml-2026-Somaskandan931"

if not os.path.exists(REPO_NAME):
    !git clone https://github.com/Somaskandan931/flyrank-ml-2026-Somaskandan931.git

# Move into the repository
%cd flyrank-ml-2026-Somaskandan931

# ============================================================
# Step 2: Verify the dataset exists
# ============================================================

print("Current Working Directory:")
print(os.getcwd())

print("\nContents of data/raw:")
!ls data/raw

# ============================================================
# Step 3: Load the dataset
# ============================================================

df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

print("\n✅ Dataset Loaded Successfully!")
print(f"Rows: {df.shape[0]}")
print(f"Columns: {df.shape[1]}")

# Show first 5 rows
df.head()

Cloning into 'flyrank-ml-2026-Somaskandan931'...
remote: Enumerating objects: 321, done.
remote: Counting objects: 100% (321/321), done.
remote: Compressing objects: 100% (123/123), done.
remote: Total 321 (delta 179), reused 304 (delta 169), pack-reused 0 (from 0)
Receiving objects: 100% (321/321), 1.88 MiB | 13.31 MiB/s, done.
Resolving deltas: 100% (179/179), done.
/content/flyrank-ml-2026-Somaskandan931
Current Working Directory:
/content/flyrank-ml-2026-Somaskandan931

Contents of data/raw:
content_refresh_anonymized.csv

✅ Dataset Loaded Successfully!
Rows: 30000
Columns: 44


,content_id,client_id,search_volume,competition,competition_level,cpc,content_type,main_intent,word_count,char_count,...,char_count_tier,ctr,avg_position,engagement_rate,scroll_rate,ai_traffic_pct,impression_tier,position_tier,trend_direction,trend_pct
0,content_304f48230142,client_f369cb89fc,10.0,0.67,HIGH,2.05,keyword article,transactional,3221.0,20457.0,...,15000-25000,0.76,10.6,5.88,4.55,0.0,good,striking,down,-41.4
1,content_a1fb4e703a9e,client_4e07408562,90.0,0.01,LOW,0.05,keyword article,informational,2481.0,15562.0,...,15000-25000,0.05,20.3,0.00,10.00,0.0,good,page_3_5,down,-57.7
2,content_9aa793d4d895,client_7f2253d7e2,0.0,0.00,LOW,0.00,keyword article,informational,3515.0,23643.0,...,15000-25000,0.09,36.5,0.00,28.57,0.0,good,page_3_5,down,-60.9
3,content_331d6c4de07b,client_19581e27de,10.0,0.00,LOW,0.00,keyword article,commercial,NaN,NaN,...,NaN,0.49,6.2,1.28,3.45,0.0,good,page_1,stable,-13.8
4,content_d99b7a2d90ca,client_3fdba35f04,0.0,0.00,LOW,0.00,keyword article,informational,2803.0,17469.0,...,15000-25000,0.13,44.0,0.00,24.29,0.0,good,page_3_5,down,-34.7


In [ ]:
# Quick look at the data

print("Rows:", len(df))
print("Columns:", len(df.columns))

print("\nUnique Clients:")
print(df["client_id"].nunique())

print("\nContent Types:")
print(df["content_type"].value_counts())

print("\nTrend Direction:")
print(df["trend_direction"].value_counts())

print("\nAverage Impressions (90d):")
print(df["impressions_90d"].mean())

print("\nAverage Sessions (90d):")
print(df["sessions_90d"].mean())

Rows: 30000
Columns: 44

Unique Clients:
32

Content Types:
content_type
keyword article       27207
feedly article         2096
comparison article      697
Name: count, dtype: int64

Trend Direction:
trend_direction
down      16262
stable     5962
up         4388
new        2236
flat       1152
Name: count, dtype: int64

Average Impressions (90d):
5200.3663

Average Sessions (90d):
37.066633333333336


## Quick Look at the Data

The starter dataset contains **30,000 content pages** across **32 anonymized clients**.

The dataset contains **44 features** describing search performance, engagement, content characteristics, and trend information. The average number of **90-day impressions is 5,200.37**, indicating that pages vary in their search visibility and performance.

The dataset also contains three content types:
- Keyword articles: **27,207**
- Feedly articles: **2,096**
- Comparison articles: **697**

Trend labels show that **16,262 pages are currently declining**, suggesting that identifying which pages should be reviewed first is an important prioritization problem.

These observations support the **Refresh / Content Opportunity Scoring** lane because manually reviewing 30,000 pages is not practical, while a data-driven ranking system can help prioritize the pages most likely to benefit from human review.

## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

In [ ]:
print("Trend Direction Distribution (%)")
print((df["trend_direction"].value_counts(normalize=True) * 100).round(2))

Trend Direction Distribution (%)
trend_direction
down      54.21
stable    19.87
up        14.63
new        7.45
flat       3.84
Name: proportion, dtype: float64


This project is intended to provide **decision support** by ranking content pages for review using observed search and engagement signals.

The recommendations are based on historical patterns in the available dataset and are intended to help human reviewers prioritize their work.

The model will identify statistical relationships within the data, but it **cannot prove** that refreshing a page will improve its performance.

This project does **not** claim to predict Google's ranking algorithm or establish causal relationships between content changes and search performance. Final decisions should always be reviewed by a human.

## Self-check

Before you submit, confirm each line honestly:

☑ Every section above is filled — markdown thinking AND the code that backs it

☑ The notebook runs top to bottom with no errors (Runtime → Run all)

☑ No client names, URLs, or private queries anywhere

☑ My claims use careful words: observed, measured, directional, decision-support

☑ Committed to my repo under work/notebooks/ — then submit your repo URL on the card.